### PR TITLE
Check Symbol instances instead of Type instances in type argument inference stacks

### LINKED ISSUE #4443: Stack overflow on recursive type alias
Code to reproduce:
```ts
//interface TreeNode {
type TreeNode = {
	name: string;
	parent: TreeNode;
}

function getNames(nodes: TreeNode[]): string[] {
	return nodes.map(s => s.name);
}
```
Compilation error:
```
(function (exports, require, module, __filename, __dirname) { /*! ************
^
RangeError: Maximum call stack size exceeded
    at instantiateAnonymousType (C:\Users\Koloto\AppData\Roaming\npm\node_modules\typescript\bin\tsc.js)
    at instantiateType (C:\Users\Koloto\AppData\Roaming\npm\node_modules\typescript\bin\tsc.js:11775:25)
    at getTypeOfInstantiatedSymbol (C:\Users\Koloto\AppData\Roaming\npm\node_modules\typescript\bin\tsc.js:10500:30)
    at getTypeOfSymbol (C:\Users\Koloto\AppData\Roaming\npm\node_modules\typescript\bin\tsc.js:10506:24)
    at inferFromProperties (C:\Users\Koloto\AppData\Roaming\npm\node_modules\typescript\bin\tsc.js:12766:40)
    at inferFromTypes (C:\Users\Koloto\AppData\Roaming\npm\node_modules\typescript\bin\tsc.js:12750:25)
    at inferFromProperties (C:\Users\Koloto\AppData\Roaming\npm\node_modules\typescript\bin\tsc.js:12766:25)
    at inferFromTypes (C:\Users\Koloto\AppData\Roaming\npm\node_modules\typescript\bin\tsc.js:12750:25)
    at inferFromProperties (C:\Users\Koloto\AppData\Roaming\npm\node_modules\typescript\bin\tsc.js:12766:25)
    at inferFromTypes (C:\Users\Koloto\AppData\Roaming\npm\node_modules\typescript\bin\tsc.js:12750:25)
```
If you replace `type` with `interface` (uncomment the first line and comment the second), the code is compiled successfully.

I use version 1.5.3, also reproduced in 1.4.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #4443.

Given

```TypeScript
type T = {
    x: T;
}

var ts: T[];
ts.map(x => x);
```

The issue is that `instantiateType` kept instantiating an anonymous type every time it dug deeper into inferring with the type literal (calling `instantiateAnonymousType` in `instantiateType`). This produces *a new type* every time, which means `isInProcess` would never succeed.

The fix is to test for the `symbol` of each type rather than the type itself.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
Pinging @ahejlsberg, @vladima, and @JsonFreeman :smile:

**Comment by ahejlsberg** [comment]
It's definitely better to check the symbol associated with type, except not all types have symbols associated with them. For example, tuple types have no associated symbols. I'd suggest you modify the logic to check if either the type objects match or the symbols associated with the types match and are non-undefined.

**Comment by ahejlsberg** [comment]
I'm not sure it entirely fixes the problem, though. For example:

```typescript
type TreeNode = {
    name: string;
    parent: TreeNode;
}
function foo<U>(x: TreeNode) { }
var n: TreeNode;
foo(n);  // Error
```

This reports an error after running into the `depth === 100` safe guard we have in `objectTypeRelatedTo`.

I'm still thinking about a better way to catch the runaway instantiation. The key issue is that `instantiateType` assumes anonymous object types never reference themselves circularly. That used to be the case before we introduced type aliases, but obviously it isn't anymore.

**Comment by DanielRosenwasser** [comment]
@ahejlsberg this is true, but that scenario isn't exacerbated by this fix. At the very least, the compiler won't crash now.

> I'd suggest you modify the logic to check if either the type objects match or the symbols associated with the types match and are non-undefined.

Would it be better to check for mutual "matchedness"? For instance, if there's a symbol, the source matches if its symbol is the same as the stack symbol, otherwise it matches if it is the same type. Then check if both the `source` and `target` match:

```TypeScript
                for (let i = 0; i < depth; i++) {
                    let sourceStackType = sourceStack[i];
                    let targetStackType = targetStack[i];

                    let sourcesMatch = sourceSymbol ?
                        sourceSymbol === sourceStackType.symbol :
                        source === sourceStackType;

                    let targetsMatch = targetSymbol ?
                        targetSymbol === targetStackType.symbol :
                        target === targetStackType;

                    if (sourcesMatch && targetsMatch) {
                        return true;
                    }
                }
```

**Comment by DanielRosenwasser** [comment]
Actually, neither of those fixes this case:

```TypeScript
type TreeNode = { leftRight: [/*left*/ TreeNode, /*right*/ TreeNode] };

function foo<U>(x: TreeNode) { }
var n: TreeNode;
foo(n);
```

**Comment by JsonFreeman** [comment]
@DanielRosenwasser Why is type argument inference diving into properties in your original repro? Seems like it should just stop immediately because it encounters a type parameter on the target side, no?

Checking for the symbol is what we switched to for isDeeplyNestedGeneric, so that it could handle infinitely expanding anonymous types. So it seems similar. For @ahejlsberg's example and your last example, I think you'd have to make a similar fix, but in objectTypeRelatedTo rather than inferFromTypes. It may involve using symbol id instead of type id in the maybe stack.

You may also have to do something similar for printing types - it may use a similar mechanism.

**Comment by JsonFreeman** [comment]
And it's true about tuple types having no symbol, but that means it's a problem in isDeeplyNestedGeneric too, because we have been comparing the symbols in that function.

File: tests/baselines/reference/typeArgumentInferenceWithClassExpression3.errors.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+
+==== tests/cases/conformance/es6/classExpressions/typeArgumentInferenceWithClassExpression3.ts (1 errors) ====
+    function foo<T>(x = class { prop: T }): T {
+        return undefined;
+    }
+    
+    foo(class { prop = "hello" }).length;
+                                  ~~~~~~
+!!! error TS2339: Property 'length' does not exist on type '{}'.
```
Comment (JsonFreeman) [inline tests/baselines/reference/typeArgumentInferenceWithClassExpression3.errors.txt]: I think this is a regression.

**Comment by JsonFreeman** [comment]
Actually, the tuples not having symbols thing is not an issue for isDeeplyNestedGeneric because we check that the type is TypeFlags.Reference or TypeFlags.Instantiated. But I don't think it would be possible to do a similar kind of check in isInProcess. You can have recursive tuple types like:
```ts
type tuple = [tuple];
```

**Comment by ahejlsberg** [comment]
I have a better fix for this issue. The core of the problem is that `instantiateType` endlessly creates new instantiations of the same type, so I decided to address that by adding a cache to the `TypeMapper` object that keeps track of instantiations of anonymous object types already created using that mapper. It fixes the original issue and passes all tests. I will put up a new PR soon.

**Comment by JsonFreeman** [comment]
Should this PR be closed?

**Comment by DanielRosenwasser** [comment]
Yup thanks for the heads up Jason

**Pull request closed without merging**